tests/unit/misc/test_cmdhistory.py

```diff
@@ -0,0 +1,185 @@
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for misc.cmdhistory.History."""
+
+import pytest
+
+from qutebrowser.misc import cmdhistory
+
+
+HISTORY = ['first', 'second', 'third', 'fourth', 'fifth']
+
+CONFIG_NOT_PRIVATE = {'general': {'private-browsing': False}}
+CONFIG_PRIVATE = {'general': {'private-browsing': True}}
+
+
+@pytest.fixture
+def hist():
+    return cmdhistory.History(history=HISTORY)
+
+
+def test_no_history():
+    hist = cmdhistory.History()
+    # .history should equal []
+    assert hist.history == []
+
+
+def test_history():
+    hist = cmdhistory.History(history=HISTORY)
+    assert hist.history == HISTORY
+
+
+@pytest.mark.parametrize('tmphist, expected', [(None, False), (HISTORY, True)])
+def test_is_browsing(hist, tmphist, expected):
+    hist._tmphist = tmphist
+    assert hist.is_browsing() == expected
+
+
+def test_start_stop(hist):
+    # We can use is_browsing() because it is tested above
+    assert not hist.is_browsing()
+    hist.start('s')
+    assert hist.is_browsing()
+    hist.stop()
+    assert not hist.is_browsing()
+
+
+def test_start_with_text(hist):
+    """Test start with given 'text'."""
+    hist.start('f')
+    assert 'first' in hist._tmphist
+    assert 'fourth' in hist._tmphist
+    assert 'second' not in hist._tmphist
+
+
+def test_start_no_text(hist):
+    """Test start with no given text."""
+    hist.start('')
+    assert list(hist._tmphist) == HISTORY
+
+
+def test_start_no_items(hist):
+    """Test start with no matching text."""
+    with pytest.raises(cmdhistory.HistoryEmptyError) as excinfo:
+        hist.start('k')
+    assert excinfo is not None
+    assert not hist._tmphist
+
+
+def test_getitem(hist):
+    """Test __getitem__."""
+    for i in range(0, len(HISTORY)):
+        assert hist[i] == HISTORY[i]
+
+
+def test_not_browsing_error(hist):
+    """Test that next/previtem throws a ValueError"""
+    with pytest.raises(ValueError) as error1:
+        hist.nextitem()
+    assert str(error1.value) == "Currently not browsing history"
+
+    with pytest.raises(ValueError) as error2:
+        hist.previtem()
+    assert str(error2.value) == "Currently not browsing history"
+
+
+def test_nextitem_single(hist, monkeypatch):
+    """Test nextitem() with valid input."""
+    hist.start('f')
+    monkeypatch.setattr(hist._tmphist, 'nextitem', lambda: 'item')
+    assert hist.nextitem() == 'item'
+
+
+def test_previtem_single(hist, monkeypatch):
+    """Test previtem() with valid input."""
+    hist.start('f')
+    monkeypatch.setattr(hist._tmphist, 'previtem', lambda: 'item')
+    assert hist.previtem() == 'item'
+
+
+def test_nextitem_previtem_chain(hist):
+    """Test a combination of nextitem and previtem statements"""
+    assert hist.start('f') == 'fifth'
+    assert hist.previtem() == 'fourth'
+    assert hist.previtem() == 'first'
+    assert hist.nextitem() == 'fourth'
+
+
+def raise_index_error(self=None):
+    raise IndexError()
```

Comment:
I'll change this myself too - the `self` parameter here is uneeded as you're assigning the function to an *instance*, so it won't get `self`. The only place where `self` is passed in is if you "assign" a function inside a *class*, e.g. by doing `def` inside `class` :wink: